nav2_pure_pursuit/src/pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,414 @@
+  node->get_parameter(plugin_name_ + ".max_angular_decel", max_angular_decel_);
+  node->get_parameter(plugin_name_ + ".lookahead_dist", lookahead_dist_);
+  node->get_parameter(plugin_name_ + ".min_lookahead_dist", min_lookahead_dist_);
+  node->get_parameter(plugin_name_ + ".max_lookahead_dist", max_lookahead_dist_);
+  node->get_parameter(plugin_name_ + ".lookahead_gain", lookahead_gain_);
+  node->get_parameter(plugin_name_ + ".max_angular_vel", max_angular_vel_);
+  double transform_tolerance;
+  node->get_parameter(plugin_name_ + ".transform_tolerance", transform_tolerance);
+  node->get_parameter(plugin_name_ + ".use_velocity_scaled_lookahead_dist",
+    use_velocity_scaled_lookahead_dist_);
+  transform_tolerance_ = tf2::durationFromSec(transform_tolerance);
+
+  global_pub_ = node->create_publisher<nav_msgs::msg::Path>("received_global_plan", 1);
+}
+
+void PurePursuitController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type pure_pursuit_controller::PurePursuitController",
+    plugin_name_.c_str());
+  global_pub_.reset();
+}
+
+void PurePursuitController::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating controller: %s of type pure_pursuit_controller::PurePursuitController\"  %s",
+    plugin_name_.c_str());
+  global_pub_->on_activate();
+}
+
+void PurePursuitController::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating controller: %s of type pure_pursuit_controller::PurePursuitController\"  %s",
+    plugin_name_.c_str());
+  global_pub_->on_deactivate();
+}
+
+double PurePursuitController::getLookAheadDistance(const geometry_msgs::msg::Twist & speed)
+{
+  // If using velocity-scaled look ahead distances, find and clamp the dist
+  // Else, use the default look ahead distance
+  double lookahead_dist = 0.0;
+  if (use_velocity_scaled_lookahead_dist_) {
+    lookahead_dist = speed.linear.x * lookahead_gain_;
+    lookahead_dist = std::clamp(lookahead_dist, min_lookahead_dist_, max_lookahead_dist_);
+  } else {
+    lookahead_dist = lookahead_dist_;
+  }
+
+  return lookahead_dist;
+}
+
+geometry_msgs::msg::TwistStamped PurePursuitController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & speed)
+{
+  // Transform path to robot base frame
+  auto transformed_plan = transformGlobalPlan(pose);
+
+  // Find look ahead distance and point on path
+  const double lookahead_dist = getLookAheadDistance(speed);
+  geometry_msgs::msg::PoseStamped carrot_pose = getLookAheadMarker(lookahead_dist, transformed_plan);
+
+  double linear_vel, angular_vel;
+
+  // Find distance^2 to look ahead point (carrot) in robot base frame
+  const double carrot_dist2 =
+    (carrot_pose.pose.position.x * carrot_pose.pose.position.x) +
+    (carrot_pose.pose.position.y * carrot_pose.pose.position.y);
+
+  // Find curvature of circular arc
+  double curvature = 0.0;
+  if (carrot_dist2 > 0.001) {
```

Comment:
Expose this as a parameter `min_curvature`?

Reply:
This isn't for tuning, this is to make sure we don't have a divide by zero situation

Reply:
Got it! Is there a reason for `0.001` or is it arbitrary? In case it is arbitrary it would be better to use this `std::numeric_limits<double>::epsilon()`.

Reply:
its 1 mm, anything bigger is reasonable (e.g. you're 1 cm away, you still want to rotate a little) but anything smaller will blow up (1 / 0.001 = 1000). Its not an exact value, but its a contextually derived order of magnitude type thing.